### PR TITLE
Microservice api page

### PR DESCRIPTION
**Pull request by MaiyoiaGamal**

### CODE REVIEW

File: app/api/microservice.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    return microServices.getAllServicesByTeamName()
+
+@router.get("/{service_id}", response_model=MicroserviceInDBBase)
+async def getOne_service(service_id: int, microServices: CRUDMicroservice = Depends(dependencies.getMicroservicesCrud)):
+    service = microServices.getByServiceIdWithTeamName(service_id)
+    if service is None:
+        raise HTTPException(status_code=404, detail="Service not found")
+    return service
+
+@router.get("/scorecards/{service_id}", response_model=MicroserviceTeamScorecardBase)
```
Comment (Passant-Abdelgalil) [inline app/api/microservice.py:26]: Please move this endpoint to a new file in the api folder called scorecards.py and do the necessary adjustments

File: app/api/microservice.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    try:
+        scorecard.getByScoreCradIds(newmicroservice.scorecardids)                                         
+    except Exception as x:
+        error_message = 'ScoreCard Id was not found'
+        print(error_message)
+        
+    created_microservice = microservice.create(MicroserviceCreate(name=newmicroservice.name,
+                                            description=newmicroservice.description,
+                                            teamId=newmicroservice.teamId,
+                                            code=newmicroservice.name.replace(" ","_")))
```
Comment (Passant-Abdelgalil) [inline app/api/microservice.py:55]: what if the newmicroservice is ` new     service  `, the code according to the current implementaion will be `-new---service-`
consider this case and update your implementation to cover it.

File: app/api/microservice.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@router.post("/create/", response_model= None)
+def create_microservice(newmicroservice: MicroserviceCreateApi, 
+                        microservice: CRUDMicroservice = Depends(dependencies.getMicroservicesCrud),
+                        teamservice: CRUDTeam = Depends(dependencies.getTeamsCrud),
+                        servicescorecard: CRUDMicroserviceScoreCard = Depends(dependencies.getMicroserviceScoreCardsCrud),
+                        scorecard: CRUDScoreCard = Depends(dependencies.getScoreCardsCrud)):
+    try:
+        teamservice.get(newmicroservice.teamid)
+        scorecard.getByScoreCradIds(newmicroservice.scorecardids)
```
Comment (Passant-Abdelgalil) [inline app/api/microservice.py:41]: no need to call this function here you already call it in a separate try except block.

File: app/api/microservice.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                        scorecard: CRUDScoreCard = Depends(dependencies.getScoreCardsCrud)):
+    try:
+        teamservice.get(newmicroservice.teamid)
+        scorecard.getByScoreCradIds(newmicroservice.scorecardids)                                         
+    except Exception as x:
+        error_message = 'Team Id was not found'
+        print(error_message)
+    
+    try:
+        scorecard.getByScoreCradIds(newmicroservice.scorecardids)
```
Comment (Passant-Abdelgalil) [inline app/api/microservice.py:47]: Try to find a way to specify which scorecard id was not found and include it in your error message.

File: app/api/microservice.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        error_message = 'Team Id was not found'
+        print(error_message)
+    
+    try:
+        scorecard.getByScoreCradIds(newmicroservice.scorecardids)                                         
+    except Exception as x:
+        error_message = 'ScoreCard Id was not found'
+        print(error_message)
+        
+    created_microservice = microservice.create(MicroserviceCreate(name=newmicroservice.name,
```
Comment (Passant-Abdelgalil) [inline app/api/microservice.py:52]: You don't check for found errors before creating the microservice. You shouldn't create the microservice If the passed values are invalid, in this case you need to respond with a HTTPException with the correct status code and the corresponding error details.

File: app/crud/microservice.py [hunk 6]
```diff
@@ -1,17 +1,72 @@
...
+        #self.teamService= CRUDTeam(db_session)
+        #self.scoreCardService= CRUDMicroserviceScoreCard(db_session)
+        
     def getByTeamId(self, teamId: str):
         return self.db_session.query(Microservice).filter(Microservice.teamId == teamId).all()
 
     def getByTeamIdAndCode(self, teamId: str, code: str):
         return self.db_session.query(Microservice).filter(Microservice.teamId == teamId, Microservice.code == code).first()
+    
+    #get all with team
```
Comment (Passant-Abdelgalil) [inline app/crud/microservice.py:26]: Please remove all commented code to keep files clean

File: app/crud/microservice.py [hunk 7]
```diff
@@ -1,17 +1,72 @@
...
+            )
+            services.append(service)
+         
+        return services
+   
+   #get one with team
+    def getByServiceIdWithTeamName(self , service_id:int):
+        result = (
+        self.db_session.query(Microservice.id, Microservice.name, Microservice.description, Microservice.code, Team.name.label("team_name"))
+        .join(Team, Microservice.teamId == Team.id)
```
Comment (Passant-Abdelgalil) [inline app/crud/microservice.py:58]: joining two tables is a costly operation, you shouldn't do it If you can obtain the information you want without joining tables.
And in this case, you can get all fields of the team without joining the two tables. You can query the Microservice table and get the record which includes the teamId, then you can use this teamId to query the Team table to get the team record. Then inject the team object in the micorservice object and return the result.
Query by Id is a fast operation, it benefits from DB tables indexing.

Please search this further and improve your code or prove that your current implementation is more efficient than the proposed one : ).

File: app/crud/microservice.py [hunk 8]
```diff
@@ -1,17 +1,72 @@
...
+        .filter(Microservice.id == service_id)
+        .first()
+        )
+        return result
+       
+#create one
+    def create_service(db: Session, service: MicroserviceCreate):
```
Comment (Passant-Abdelgalil) [inline app/crud/microservice.py:65]: this function is useless, there's a function already to create a new record in a DB table called `create` and it exists in the parent class CRUDBase and in the current child class due to inheritance.
Please remove this function and start using the `create` function.

File: app/crud/microserviceTeamScorecard.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
+import sqlalchemy
```
Comment (Passant-Abdelgalil) [inline app/crud/microserviceTeamScorecard.py:1]: This file is duplicated with a different name `microserviceTeamScoreCard.py`. Please remove one of them and do the necessary adjustments.

File: app/crud/microserviceTeamScoreCard.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            raise Exception(f"Microservice with id {service_id} not found")
+        
+        # Fetch the associated team
+        teamobject = self.teamService.get(microservice.teamId)  # Adjusted to match your model's attribute
+
+        # Fetch the scorecards associated with the microservice
+        scorecardIds = self.scoreCardService.getByServiceId(microservice.id)
+        scorecards = []
+        
+        for sc_id in scorecardIds:
```
Comment (Passant-Abdelgalil) [inline app/crud/microserviceTeamScoreCard.py:35]: I believe you can write a custom function in the CRUDScorecard class, the function can write a custom sql filter query such that given a list of IDs it filters table records with id equal to one of the values in the provided list.
This way we query the table only once instead of repeating the query as many times as the length of the scorecardIds.

Please search this further and improve your code or prove this your current implementaion is more efficient : )

Comment (Passant-Abdelgalil) [reply app/crud/microserviceTeamScoreCard.py:35]: I can see that there's already a function called `getByScoreCradIds`  in the CRUDScoreCard class why don't you use it ? : )

File: app/crud/microserviceTeamScoreCard.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+from ..schemas import MicroserviceTeamScorecardBase,MicroserviceTeamScorecardCreate,MicroserviceTeamScorecardUpdate, team, scoreCard
+from .base import CRUDBase
+from typing import List
+from sqlalchemy.sql import func
+from starlette.exceptions import HTTPException
+from . import CRUDTeam, CRUDMicroserviceScoreCard, CRUDScoreCard
+
+
+
+class CRUDMicroserviceTeamScorecard(CRUDBase[MicroserviceTeamScorecardBase, MicroserviceTeamScorecardCreate, MicroserviceTeamScorecardUpdate ]):
```
Comment (Passant-Abdelgalil) [inline app/crud/microserviceTeamScoreCard.py:13]: This is not a class that directly operate on a single DB table. You shouldn't make it inherit from the CRUDBase class.
ONLY classes that directly operate on a SINGLE  Database table can inherit form CRUDBase class.

This is a utility class more than a CRUD class, And I believe there's no need to make it a separate class. You only use  `getByServiceIdWithTeamAndSCOREDetails` method. Why don't you create a new file inside a folder called utils or helpers for example and move this `getByServiceIdWithTeamAndSCOREDetails` function in that file.

File: app/models/scorecard.py [hunk 12]
```diff
@@ -1,12 +1,11 @@
...
 from sqlalchemy import Column, Integer, String
 
-from ...database import Base
-
 
 class Scorecard(Base):
     id = Column(Integer, primary_key=True, index=True)
-    name = Column(String(50), nullable=False)
-    description = Column(String(200), nullable=False)
+    name = Column(String, nullable=False)
```
Comment (Passant-Abdelgalil) [inline app/models/scorecard.py:7]: Why did you remove the limit on the allowed number of characters for each column ?

File: app/models/team.py [hunk 13]
```diff
@@ -10,7 +10,7 @@ class Team(Base):
     id = Column(UUID(as_uuid=True), primary_key=True, index=True, default=uuid.uuid4)
     name = Column(String, nullable=False)
     token = Column(String, nullable=False)
-    microservices = relationship("Microservice", cascade="all, delete-orphan", primaryjoin="Microservice.teamId == Team.id")
-
+    #microservices = relationship("Microservice", cascade="all, delete-orphan", primaryjoin="Microservice.teamId == Team.id")
+    microservices = relationship("Microservice", back_populates="team")
```
Comment (Passant-Abdelgalil) [inline app/models/team.py:14]: Why did you do this change ?

File: app/schemas/microservice.py [hunk 14]
```diff
@@ -17,25 +16,37 @@ class MicroserviceCreate(MicroserviceBase):
...
 class MicroserviceInDBBase(MicroserviceBase):
     id: int
     name: str
-    code: str
+    code: Optional[str] = None   
     description: str
-    teamId: uuid.UUID
+    #teamId: uuid.UUID
+    #add this line
+    team_name: Optional[str] =None
```
Comment (Passant-Abdelgalil) [inline app/schemas/microservice.py:44]: `MicroserviceInDBBase` should include properties that already exist in the DB table.
If you want to include other properties, create a new class.

You shouldn't change the current definitions unless you are 100% sure that your changes don't break anything in the code.

File: app/schemas/microserviceScoreCard.py [hunk 15]
```diff
@@ -1,13 +1,18 @@
...
+from pydantic import BaseModel 
 
 
 # Shared properties
 class MicroserviceScoreCardBase(BaseModel):
     microserviceId: Optional[int] = None
     scoreCardId: Optional[int] = None
 
+#added 
+class ScoreCardBase(BaseModel):
```
Comment (Passant-Abdelgalil) [inline app/schemas/microserviceScoreCard.py:12]: Why is this class needed? Why don't you use the `ScoreCardInDBBase` class defined in the scoreCard.py file ?

File: app/schemas/microserviceTeamScorecard.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
+from typing import Optional
```
Comment (Passant-Abdelgalil) [inline app/schemas/microserviceTeamScorecard.py:1]: Again, this file is a duplicate of microserviceTeamScoreCard.py.  Remove one of them.

File: app/api/microservice.py (hunk 1 above)
Comment (MaiyoiaGamal) [reply app/api/microservice.py:26]: I believe this is a one of Microservice operation to return all service detail , so i will update the path name 
https://github.com/tactful-ai/microcentral/pull/43/files#diff-8dc9ecf2eabdfea0bed728acf1163c7da8af51a22d552d1e12c908852128143dR26

File: app/api/microservice.py (hunk 2 above)
Comment (MaiyoiaGamal) [reply app/api/microservice.py:55]: Doen

File: app/api/microservice.py (hunk 3 above)
Comment (MaiyoiaGamal) [reply app/api/microservice.py:41]: done

File: app/api/microservice.py (hunk 4 above)
Comment (MaiyoiaGamal) [reply app/api/microservice.py:47]: done

File: app/api/microservice.py (hunk 5 above)
Comment (MaiyoiaGamal) [reply app/api/microservice.py:52]: done
